Add Styleguide (#7515)
ComplianceAsCode/content: docs/manual/developer/04_style_guide.md

# Compliance as Code Content Style Guide

## Pull Requests
* Should follow the template
* Shall remove the sample text from the template
* Shall not have merge commits

### Before Merging 

> * Must have released associated

Reviewer: ```suggestion
* The milestone must be set correctly.
```

* Must have the correct labels
* Should be assigned to the reviewers

> 

Reviewer: ```suggestion
* There are no merge commits - if that's the case , they have to be taken out by rebasing.

```

* `template`

### Group
This section describes the styleguide around the `group.yml` files. 
All the above [YAML](manual/developer/04_style_guide:yaml) rules apply.
A group should only contain rules that effect the same software or service.

#### Sections
Group sections should be in the following order, if present.
* `documentation_complete`
* `title`
  * Must be in [Title case](https://en.wikipedia.org/wiki/Title_case)
* `description` (HTML-Like)

### Benchmark
This section describes the styleguide around the `benchmark.yml` files. 
All the above [YAML](manual/developer/04_style_guide:yaml) rules apply.

#### Sections
Benchmark sections should be in the following order, if present.
* `documentation_complete`
* `status`
* `title`
  * Must be in [Title case](https://en.wikipedia.org/wiki/Title_case)
* `description` (HTML-Like)
* `notice` (HTML-Like)
* `front-matter` (HTML-Like)
* `rear-matter` (HTML-Like)
* `version`

## Remediation
All remediations should have the following header with the correct 
```bash
# platform = multi_platform_all
# reboot = false
# strategy = enable
# complexity = low
# disruption = low
```

### Bash
* Should use marcos instead shared functions

> * 

Reviewer: ```suggestion
* Use 4-space indentation
* Put `do` or `then` on the same line as `for` or `if` respectively, e.g. `for file in *; do`

```

### Ansible
* Shall follow all the rules in the [YAML](manual/developer/04_style_guide:yaml) section
* Should prefer using Ansible modules over just calling system commands

## XML
* Shall use the `.xml` for the file extension
* Must be able to be parsed by Python's XML parser

### OVAL
* The `id` attribute of `<definition>` should be `{{{ rule_id }}}`
* The elements should be in the following order:
  * `def-group`
    * `definition`
      * `metadata`
        * Should be defined by the`oval_metadata` marco
      * `criteria`
    * * Like OVAL `test`, `object`, and `state` should have the same name after there respective prefix.
    * Like OVAL `test`, `object`, and `state` should group together, if there are many tests the order below should be repeated for each group.
      * `test`
      * `object`
      * `state`

### Test elements
* `id` should start with `test_`

### Object elements
* `id` should start with `obj_`

### State elements
* `id` should start with `state_`

## Tests
* Should always do something, even if testing default behavior
* Should test one change

## Markdown

> * One sentence per line

Reviewer: ```suggestion
* [One sentence per line](https://asciidoctor.org/docs/asciidoc-recommended-practices/#one-sentence-per-line)
```

# Compliance as Code Content Style Guide

## Pull Requests
* Should follow the template
* Shall remove the sample text from the template
* Shall not have merge commits, they should have to be taken out by rebasing.

### Before Merging 
* The milestone must be set correctly
* Must have the correct labels
* Should be assigned to the reviewers

### Merging
* Should use the merge commit method

## All files
* Shall use UNIX style line endings

> * Shall have one blank line at the end of the file

Reviewer 2: Did you mean a newline at the end of the file, instead of a blank line?

* Methods should be defined before they called
* The files in the build system shall be Python 2.7 and Python 3 compatible
* Utilities may only be compatible Python 3
* Shall use the `.py` for the file extension
* Use 4-space indentation

## YAML
* All YAML files shall use 4 spaces inplace of tabs.
* Maximum line length is 99 characters
* Must be able to be parsed with PyYAML
* Shall use the `.yml` vs `.yaml` for the file extension
* Shall have one blank line between sections

### HTML Like Fields
In the sections below marked with an `(HTML Like)` means that a limited number of HTML elements are supported in these fields.

We support the following elements:
* `tt` - inline code blocks
* `b` - bold face
* `i` - italics 
* `pre` - code block
* `br` - line break

### Rule
This section describes the styleguide around the `rule.yml` files. 
All the above [YAML](manual/developer/04_style_guide:yaml) rules apply.
A rule should only address one configuration item change.

#### Sections
Rules sections should be in the following order, if present.
* `documentation_complete`
* `prodtype`
  * Comma separated list
  * No spaces between items
  * Items must be in alphabetical order
* `title`

>   * Must be One Line 

Reviewer 2: ```suggestion
  * Must be one line 
```

  * Must be in [Title case](https://en.wikipedia.org/wiki/Title_case)
  * Must be short and descriptive
  * Must roughly match the path the `rule.yml` is in
* `description` (HTML Like)
* `rationale` (HTML Like)
* `severity`
* `identifiers`
  * Keys must be in alphabetical order
* `references`
  * Keys must be in alphabetical order
* `platforms`
* `ocil_clause`
* `ocil` (HTML Like)
* `warnings`
  * All subsections are HTML-Like
  * If defined must have at least one of the following sub-sections:
    * `general`
    * `dependency`
    * `performance`
    * `management`
    * `functionality`
* `conflicts`
  * Must be a valid rule id
* `requires`
  * * Must be a valid rule id
* `template`

### Group
This section describes the styleguide around the `group.yml` files. 
All the above [YAML](manual/developer/04_style_guide:yaml) rules apply.
A group should only contain rules that effect the same software or service.

#### Sections

> Group sections should be in the following order, if present.

Reviewer 2: A group can have `platforms` key too.

* `documentation_complete`
* `title`
  * Must be in [Title case](https://en.wikipedia.org/wiki/Title_case)
* `description` (HTML-Like)

### Benchmark
This section describes the styleguide around the `benchmark.yml` files. 
All the above [YAML](manual/developer/04_style_guide:yaml) rules apply.

#### Sections
Benchmark sections should be in the following order, if present.
* `documentation_complete`
* `status`
* `title`
  * Must be in [Title case](https://en.wikipedia.org/wiki/Title_case)
* `description` (HTML-Like)
* `notice` (HTML-Like)
* `front-matter` (HTML-Like)
* `rear-matter` (HTML-Like)
* `version`

## Remediation

> All remediations should have the following header with the correct 

Reviewer 2: ```suggestion
All remediations should have the following header with the appropriate values
```

```bash
# platform = multi_platform_all
# reboot = false
# strategy = enable
# complexity = low
# disruption = low
```

### Bash

> * Should use marcos instead shared functions

Reviewer 2: ```suggestion
* Should use macros instead shared functions
```

* Should prefer using Ansible modules over just calling system commands

## XML
* Shall use the `.xml` for the file extension
* Must be able to be parsed by Python's XML parser

### OVAL
* The `id` attribute of `<definition>` should be `{{{ rule_id }}}`
* The elements should be in the following order:
  * `def-group`
    * `definition`
      * `metadata`
        * Should be defined by the`oval_metadata` marco
      * `criteria`
    * * Like OVAL `test`, `object`, and `state` should have the same name after there respective prefix.
    * Like OVAL `test`, `object`, and `state` should group together, if there are many tests the order below should be repeated for each group.
      * `test`
      * `object`
      * `state`

### Test elements
* `id` should start with `test_`

### Object elements
* `id` should start with `obj_`

### State elements
* `id` should start with `state_`

## Tests
* Should always do something, even if testing default behavior
* Should test one change

## Markdown
* [One sentence per line](https://asciidoctor.org/docs/asciidoc-recommended-practices/#one-sentence-per-line)

### Headings

> * Must have one blank file above headings

Reviewer 2: Typo, I guess?
```suggestion
* Must have one blank line above headings
```

> * Must be in [Title case](https://en.wikipedia.org/wiki/Title_case)

Reviewer 2: Should this apply to all heading levels?
I see, for example,  in line 165 `### State elements`, should it be `### State Elements`?

* Should prefer using Ansible modules over just calling system commands
* Shall be written to pass `ansible-lint`

### Kubernetes
* Shall follow all the rules in the [YAML](manual/developer/04_style_guide:yaml) section

## XML
* Shall use the `.xml` for the file extension
* Must be able to be parsed by Python's XML parser
* Shall use 4-space indentation

### OVAL
* The `id` attribute of `<definition>` should be `{{{ rule_id }}}`
* The elements should be in the following order:
  * `def-group`
    * `definition`
      * `metadata`
        * Should be defined by the`oval_metadata` macro
      * `criteria`
    * Like OVAL `test`, `object`, and `state` should have the same name after their respective prefix.
    * Like OVAL `test`, `object`, and `state` should group together, if there are many tests the order below should be repeated for each group.
      * `test`
      * `object`
      * `state`

### Test Elements
* `id` should start with `test_`

### Object Elements
* `id` should start with `obj_`

### State Elements
* `id` should start with `state_`

## Tests
* Should always do something, even if testing default behavior
* Should test one change

> * Shall use the `.sh` for the file extension

Reviewer 3: ```suggestion
* Shall use the `.sh` for the file extension
* Should use the `#!/bin/bash` shebang at the first line
```

# ComplianceAsCode Style Guide

## Pull Requests
* Should follow the template
* Shall remove the sample text from the template pull request
* Shall not have merge commits, they should have to be taken out by rebasing

### Before Merging 
* The milestone must be set correctly
* Must have the correct labels

> * Should be assigned to the reviewers

Author: Reason why I put in the Before Merging section. My goal of this section is to be checklist for the maintainer merging the PR. Maybe I need to add something to make it clearer.

### Rule
This section describes the styleguide around the `rule.yml` files. 
All the above [YAML](manual/developer/04_style_guide:yaml) rules apply.
A rule should only address one configuration item change.
A variable should be when a configuration change can be multiple different values.

#### Rule Sections
Rules sections must be in the following order, if present.
* `documentation_complete`
* `prodtype`
  * Comma separated list
  * No spaces between items
  * Items must be in alphabetical order
* `title`
  * Must be one line 
  * Must be in [Title case](https://en.wikipedia.org/wiki/Title_case)
  * Must be short and descriptive
  * Must roughly match the directory name the `rule.yml` is in
* `description` (HTML Like)
* `rationale` (HTML Like)
* `severity`
* `identifiers`
  * Keys must be in alphabetical order
* `references`
  * Keys must be in alphabetical order
* `platforms`
* `ocil_clause`
* `ocil` (HTML Like)
* `warnings`
  * All subsections are HTML-Like
  * If defined must have at least one of the following sub-sections:
    * `general`
    * `dependency`
    * `performance`
    * `management`
    * `functionality`
* `conflicts`
  * Must be a valid rule id
* `requires`

>   * * Must be a valid rule id

Reviewer 4: nit: Double star I think is a typo?

  * Should be the same as the standard
* `source`
  * URL to the standard
* `controls`
  * `id`
  * `levels`
    * Should be in lowercase
    * Must have words separated by an underscore
    * Shall follow the standard
  * `title`
    * Shall be one line
  * `status`
  * `notes`
    * Must be a block
  * `rules`
    * Must be a valid rule id

## Remediation
All remediations should have the following header with the appropriate values 
```bash
# platform = multi_platform_all
# reboot = false
# strategy = enable
# complexity = low
# disruption = low
```

### Bash
* Should use macros instead shared functions
* Shall use 4-space indentation
* Put `do` or `then` on the same line as `for` or `if` respectively, e.g. `for file in *; do`

### Ansible
* Shall follow all the rules in the [YAML](manual/developer/04_style_guide:yaml) section
* Should prefer using Ansible modules over just calling system commands
* Shall be written to pass `ansible-lint`

### Kubernetes
* Shall follow all the rules in the [YAML](manual/developer/04_style_guide:yaml) section

## XML
* Shall use the `.xml` for the file extension
* Must be able to be parsed by Python's XML parser
* Shall use 4-space indentation

> ### OVAL

Reviewer 4: Do we want to add something about OVAL comments here? IMO they're really helpful and while mandated in certain places, aren't mandated everywhere and thus we should encourage their use and clarity?

* The `id` attribute of `<definition>` should be `{{{ rule_id }}}`
* The elements should be in the following order:
  * `def-group`
    * `definition`
      * `metadata`
        * Should be defined by the`oval_metadata` macro
      * `criteria`
    * Like OVAL `test`, `object`, and `state` should have the same name after their respective prefix.
    * Like OVAL `test`, `object`, and `state` should group together, if there are many tests the order below should be repeated for each group.
      * `test`
      * `object`
      * `state`

> ### Test Elements

Reviewer 4: Should these be indented to clarify they're part of OVAL?

* `id` should start with `test_`

### Object Elements
* `id` should start with `obj_`

### State Elements
* `id` should start with `state_`

## Tests
* Shall always do something, even if testing default behavior
* Shall test one change
* Shall use the `.sh` for the file extension
* Shall use the `#!/bin/bash` shebang at the first line

## Markdown
* [One sentence per line](https://asciidoctor.org/docs/asciidoc-recommended-practices/#one-sentence-per-line)

> * Shall use the `.md` for the file extension

Reviewer 3: The link is for asciidoc which is a different format than markdown

Author: This is true, but the ideas are similar @matejak suggested this link.
